rapidsai/cudf
# [WIP] Updating CMakeLists.txt to use CPM for Package Management and / take advantage of cmake 3.18 support.

In an attempt to wrangle `RAPIDS` `CMakeLists.txt` I've adopted the use of `CPM` as a package manager and audited some other issues. This PR is not meant to dictate where to go with `cmake` but just an attempt to solve a few problems and get some discussion going about solving problems going forward. Just to try to streamline that discussion I've outlined what I'm doing and why below as well as some excerpts from the `CPM` documentation. 

My tear down and rebuild process for `cudf` was all in the context of `rapids-compose` so I could easily destroy `ccache` and `conda` to verify multiple states. I do not think this PR fully reflects the MVP to merge this pattern. 

I also did a treatment using the package manager `hunter` instead of `CPM` (Draft PR #6026) . I found that `hunter` was more ops focused and had a pattern focused on building a repository of built ccache binaries. `CPM` is a much simpler project and seems to fit our goals better.

CMakeTests.txt Outline
==================
I drew this little thing to give an overview of the existing `CMakeLists.txt`, it's functions, and what is modified when upgrading to cmake 3.18 and using CPM. I numbered the sections to make it easier to discuss in this PR.

![CUDF-CMAKE-Upgrade-Overview](https://user-images.githubusercontent.com/100833/91334494-ab19eb00-e794-11ea-8a66-c2f55958e222.png)

Problems and Goals
================== 

Problems
--------
* Our `cmake` (in all `RAPIDS` projects) has accumulated some neglected copypasta that bogs down production and steals productivity 
* We cannot upgrade to the newest `cmake` (3.18) which has very useful helpers for `CUDA` support because of issues relating to our complicated environment flow
* Not having a simple, standards driven `cmake` causes repetition of bad but functional tooling
* `cmake` is an afterthought for many developers slowing down production when `cmake` ontology is not clear
* New `cmake 3.18` `CUDA` features are broken when we do our `conda` environment steps in our `CMakeList.txt`
  - Do we still need our `CUDA` architecture detection?
  - In some environments [`rapids-compose`] does our current `conda` environment break the new (`cmake 3.18`) `nvcc` / `cuda` detection?
    * `/conda/cuda_10.2/envs/rapids/compiler_compat/ld: cannot find -lcudf`

Goals
-----
* Make `RAPIDS` `cmake` compatible with `v3.18` to utilize new native `CUDA` features
* Do not build a new `cmake` package management paradigm using `CPM`. 
* Decrease `RAPIDS` install conflicts related to existing system installed packages
* Pair down redundant and complicated environment vars, includes, and linked directories.
  - Standardize target linking to enable using `PUBLIC` and `PRIVATE` instead of passing around package name variables.
* Speed up build times
* Increase resilience and predictability to dependency management
* A single source of truth for future configurations in `RAPIDS` projects
* Easy incremental integration without disrupting development
* Easy CI/CD integration
* Easy improvements, enhancements, and ability to refactor for `cmake` configuration for all rapids projects
* Reduce malignant duplication between `RAPIDS` projects `cmake` configurations


Overview of Changes
===================
* Adding `CPM` as the `cmake` package manager.
  - https://github.com/TheLartians/CPM.cmake
* Simplifying `zlib`, `boost-filesystem`, `gtest`, and `benchmarks` management via `CPM`
* Adding `gtest` and `benchmarks` to the singluar `target_link_libraries`
  - I am unsure about the universality of `${<package>_LIBRARIES}` usage but hopeful that it can work as a model for single point collection
* [WIP] Using new `FindCUDAToolkit` features of `cmake 3.18` to simplify GPU architecture detection and loading of `cudart` packages.
  - Currently the detection of the package is failing, though it seems to be able to identify the installed `CUDA` version.
* [TODO] Build `arrow` with `CPM` falling back to system if correct options are flagged, if not build from source.

I have not directly addressed but am aware of @jdye64 's outline for changes to our cmake configuration. Some of his issues are partially addressed and I believe we can encapsulate the issues he's brought up in this push to adopt 3.18 and CPM.

Ask @jdye64 for a link if you're curious about the document. Issues #5940 and #5569 are two thirds of it.

Testing Procedure
=================
To properly test this you should use Paul Taylor's [@trxcllnt] `rapids-compose` project [github.com/trxcllnt/rapids-compose] and not your system directly. This will make it easier to modify environment to understand all the variable involved and also not waste too much time and possibly upset your normal development environment.

About CPM
==========

Comparison to pure FetchContent / ExternalProject
--------------------------------------------------------------------
https://github.com/TheLartians/CPM.cmake#comparison-to-pure-fetchcontent--externalproject

CPM.cmake is a wrapper for CMake's FetchContent module and adds a number of features that turn it into a useful dependency manager. The most notable features are:

* A simpler to use API
* Version checking: CPM.cmake will check the version number of any added dependency and omit a warning if another dependency requires a more recent version.
* Options: any Options passed to a dependency are stored and compared on later use, so if another dependency tries to add an existing dependency with incompatible options a warning will be emitted to the user.
* Offline builds: CPM.cmake will override CMake's download and update commands, which allows new builds to be configured while offline if all dependencies are available locally.
* Automatic shallow clone: if a version tag (e.g. v2.2.0) is provided and CPM_SOURCE_CACHE is used, CPM.cmake will perform a shallow clone of the dependency, which should be significantly faster while using less storage than a full clone.
* Overridable: all CPMAddPackage can be configured to use find_package by setting a CMake flag, making it easy to integrate into projects that may require local versioning through the system's package manager.
* Package lock files for easier transitive dependency management.
* Dependencies can be overridden per-build using CMake CLI parameters.
* ExternalProject works similarly as FetchContent, however waits with adding dependencies until build time. This has a quite a few disadvantages, especially as it makes using custom toolchains / cross-compiling very difficult and can lead to problems with nested dependencies.
